zilliqa/src/inspector.rs

```diff
@@ -350,3 +356,91 @@ impl ScillaInspector for OtterscanOperationInspector {
 }
 
 impl ScillaInspector for TracingInspector {}
+
+impl<I: Inspector<PendingState> + ScillaInspector> ScillaInspector for ZQ2Inspector<'_, I> {}
+
+pub struct ZQ2Inspector<'a, I: Inspector<PendingState> + ScillaInspector> {
+    inner: &'a mut I,
+}
+
+impl<'a, I: Inspector<PendingState> + ScillaInspector> ZQ2Inspector<'a, I> {
+    pub fn new(inner: &'a mut I) -> Self {
+        Self { inner }
+    }
+}
+
+impl<I: Inspector<PendingState> + ScillaInspector> Inspector<PendingState> for ZQ2Inspector<'_, I> {
+    fn initialize_interp(
+        &mut self,
+        interp: &mut Interpreter,
+        context: &mut EvmContext<PendingState>,
+    ) {
+        context.db.gas_left = interp.gas;
+        self.inner.initialize_interp(interp, context);
+    }
+    fn step(&mut self, interp: &mut Interpreter, context: &mut EvmContext<PendingState>) {
+        context.db.gas_left = interp.gas;
+        self.inner.step(interp, context);
+    }
+
+    fn step_end(&mut self, interp: &mut Interpreter, context: &mut EvmContext<PendingState>) {
+        self.inner.step_end(interp, context);
+    }
+
+    fn log(&mut self, interp: &mut Interpreter, context: &mut EvmContext<PendingState>, log: &Log) {
+        self.inner.log(interp, context, log);
+    }
+    fn call(
+        &mut self,
+        context: &mut EvmContext<PendingState>,
+        inputs: &mut CallInputs,
+    ) -> Option<CallOutcome> {
+        Inspector::call(&mut self.inner, context, inputs)
```

Comment:
If these methods are not changing the default implementation from the `Inspector` trait, we should not list them here.